src/metabase/query_processor/middleware/optimize_temporal_filters.clj

```diff
@@ -3,65 +3,104 @@
   (:require
    [clojure.walk :as walk]
+   [metabase.lib.hierarchy :as lib.hierarchy]
    [metabase.mbql.util :as mbql.u]
    [metabase.util :as u]
    [metabase.util.date-2 :as u.date]
-   [metabase.util.i18n :refer [trs]]
    [metabase.util.log :as log]
    [metabase.util.malli :as mu]
    [metabase.util.malli.schema :as ms]))
 
 (def ^:private optimizable-units
   #{:second :minute :hour :day :week :month :quarter :year})
 
-(defn- temporal-unit [field]
-  (mbql.u/match-one field [:field _ (opts :guard :temporal-unit)] (:temporal-unit opts)))
-
-(defn- optimizable-field? [field]
-  (mbql.u/match-one field
-    [:field _ (_ :guard (comp optimizable-units :temporal-unit))]))
-
-(defmulti ^:private can-optimize-filter?
-  mbql.u/dispatch-by-clause-name-or-class)
-
-(defn- optimizable-temporal-value?
-  "Can `temporal-value` clause can be optimized?"
-  [temporal-value]
-  (mbql.u/match-one temporal-value
-    [:relative-datetime (_ :guard #{0 :current})]
-    true
-
-    [(_ :guard #{:absolute-datetime :relative-datetime}) _ (unit :guard optimizable-units)]
-    true))
-
-(defn- field-and-temporal-value-have-compatible-units?
-  "Do datetime `field` clause and `temporal-value` clause have 'compatible' units that mean we'll be able to optimize
-  the filter clause they're in?"
-  [field temporal-value]
-  (mbql.u/match-one temporal-value
-    [:relative-datetime (_ :guard #{0 :current})]
-    true
-
-    [(_ :guard #{:absolute-datetime :relative-datetime}) _ (unit :guard optimizable-units)]
-    (= (temporal-unit field) unit)))
-
-(defmethod can-optimize-filter? :default
-  [filter-clause]
-  (mbql.u/match-one filter-clause
-    [_
-     (field :guard optimizable-field?)
-     (temporal-value :guard optimizable-temporal-value?)]
-    (field-and-temporal-value-have-compatible-units? field temporal-value)))
-
-(defmethod can-optimize-filter? :between
-  [filter-clause]
-  (mbql.u/match-one filter-clause
-    [_
-     (field :guard optimizable-field?)
-     (temporal-value-1 :guard optimizable-temporal-value?)
-     (temporal-value-2 :guard optimizable-temporal-value?)]
-    (and (field-and-temporal-value-have-compatible-units? field temporal-value-1)
-         (field-and-temporal-value-have-compatible-units? field temporal-value-2))))
+(defmulti ^:private temporal-unit
+  {:arglists '([expression])}
+  mbql.u/dispatch-by-clause-name-or-class
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod temporal-unit :default
+  [_expr]
+  nil)
+
+(defmethod temporal-unit :field
+  [[_field _id-or-name opts]]
+  (:temporal-unit opts))
+
+(defmethod temporal-unit :relative-datetime
+  [[_relative-datetime _n unit]]
+  unit)
+
+(defmethod temporal-unit :absolute-datetime
+  [[_absolute-datetime _t unit]]
+  unit)
+
+(doseq [tag [:+ :-]]
+  (lib.hierarchy/derive tag ::arithmetic-expr))
+
+(defmethod temporal-unit ::arithmetic-expr
+  [[_tag & args]]
+  (some temporal-unit args))
+
+(defmethod temporal-unit :+
+  [[_+ & args]]
+  (some temporal-unit args))
```

Comment:
Isn't this redundant with the `::arithmetic-expr` version? (Does `derive` work on un-namespaced keywords?)